Create before_after_filesystem_snapshot.py

IshaVDave opened the pull request:
This PR aims to resolve a sub-task of Issue #6 by introducing a WIP feature to the layout-web-tool which has 2 functions, corresponding tests and sets up Travis for this repository. The first function 'snapshot' returns which files were unchanged, modified, added, or removed from an input dictionary (before_dict) and an output dictionary (after_dict). The second function 'generate_artifact_rules' generates artifact rules for the previous function.

IshaVDave committed:
Create before_after_filesystem_snapshot.py

Introduces a function 'snapshot' to before_after_filesystem_snapshot.py,
that returns which files were unchanged, modified, added, or removed from an
input dictionary (before_dict) and an output dictionary (after_dict).

lukpueh commented on before_after_filesystem_snapshot.py:
```diff
@@ -0,0 +1,45 @@
+def before_after_snapshot(before_dict, after_dict):
+  '''before_after_snapshot is a simple function that outputs which files were
+    unchanged, modified, added or removed after a step was carried out.'''
```
I generally recommend to stay within the scope of the function, when writing a docstring. A reader might not know what "a step was carried out means", and also does not need to. I, as a user of this function, would rather like to know what the arguments need to look like. :)

That said, you can still mention in-toto steps to add some context or an example use case.

lukpueh commented on before_after_filesystem_snapshot.py:
```diff
@@ -0,0 +1,45 @@
+'one.tgz': '1234567890abcdef',
+'foo/two.tgz': '0000001111112222',
+'three.txt': '1111222233334444',
+'bar/bat/four.tgz': '6677889900112233'
```
Please use the same indentation as you do for the `after` dict.

lukpueh commented on before_after_filesystem_snapshot.py:
```diff
@@ -0,0 +1,45 @@
+  print('unchanged = ', str(unchanged_files))
+  print('modified = ', str(modified_files))
+  print('added = ', str(added_files))
+  print('removed = ', str(removed_files))
```
The function becomes more useful if it returns the lists. Then the caller can decide if they want to print or further process them. 

 (hint: Python can have multiple return values)

lukpueh reviewed (commented):
Impeccable work, @IshaDave! 🎉  I left a few high-level recommendations inline, but this looks really good. As a next step, it would be nice if you created a test for your function. You can use the sample dictionaries for that.

lukpueh commented:
We don't have tests in this repo yet. I suggest you create a `tests` directory with a `test_before_after_filesystem_snapshot.py` file. In in-toto, tuf and securesystemslib we use Python's built-in `unittest` framework. Since this is a fresh start, feel free to toy around with alternative tools like `pytest`.

IshaVDave commented:
Thank you so much for your comments @lukpueh! I resolved the concerns and I'm working on writing a test now. I think I am going to try the unittest framework itself, because I have been wanting to use it, for a while. :)

IshaVDave committed:
Create tests directory and add some tests.

Initiates the test directory for this repository, and adds tests
for before_after_filesystem_snapshot.py.

adityasaky commented on before_after_filesystem_snapshot.py:
```diff
@@ -0,0 +1,29 @@
+      added_files.append(key)
+
+  # Returning the snapshot of the new file system
+  return'unchanged = ' + str(sorted(unchanged_files)) + '\n' + \
```
Can we just return the four lists here instead of casting to a string?

adityasaky commented:
Can we introduce a travis / appveyor config file in this PR to run the tests?

IshaVDave commented:
Okay, will do!

IshaVDave committed:
Create .travis.yml.

Sets up Travis CI for this repository.

lukpueh commented on tests/test_before_after_filesystem_snapshot.py:
```diff
@@ -0,0 +1,113 @@
+      'bar/bat/four.tgz': '6677889900112233'
+    }
+
+    snapshot = before_after_filesystem_snapshot.snapshot(Test_before_after_filesystem_snapshot.before, after)
```
Hint: You can also use `self.before` (makes it a bit shorter).

lukpueh commented on tests/test_before_after_filesystem_snapshot.py:
```diff
@@ -0,0 +1,113 @@
+      'three.txt'], [], [], []))
+
+
+  def test_empty_filesystem_snapshot(self):
```
Minor nit: I'd call this something like `test_removed..`

lukpueh commented on before_after_filesystem_snapshot.py:
```diff
@@ -0,0 +1,64 @@
+    artifact_rules['expected_products'].append(["DELETE", file])
+  artifact_rules['expected_products'].append(["DISALLOW", "*"])
+
+  return sorted(artifact_rules)
```
I doubt that this is what you want. It will return a sorted list of the keys of the `artifact_rules` dict, i.e. `['expected_materials', 'expected_products']`. 

I'd probably define the two rule lists as variables outside of a dict and then just...
```python
return {
  'expected_materials': expected_materials,
  'expected_products': expected_products
}
```

(this would also make the lines above a tad shorter. :)

lukpueh commented on before_after_filesystem_snapshot.py:
```diff
@@ -0,0 +1,64 @@
+  }
+
+  # missing rules for MATCH since we don't have the information of the material from 
+  # the previous step
```
❤️  for adding this comment. This is actually the interesting part, that is generating match rules that are a bit smarter than just matching everything from the previous step (which is what I did in `create_layout.py`).

lukpueh replied:
Looking forward to seeing more on this front! :)

lukpueh commented on before_after_filesystem_snapshot.py:
```diff
@@ -0,0 +1,64 @@
+    artifact_rules['expected_products'].append(["CREATE", file])
+  for file in snapshot[3]:
+    # removed files
+    artifact_rules['expected_products'].append(["DELETE", file])
```
Note that the delete rule does not have an effect if it is a product rule. This is because product rules can only consume artifacts from the product queue, but a deleted artifact won't show up in the product queue, it only shows up in the materials queue, so it must be consumed there.

For the create rule it's the other way around.

(cross-referencing https://github.com/in-toto/docs/issues/4)

lukpueh replied:
Let me know if this isn't clear.

SolidifiedRay replied:
Thanks a lot for the review! @lukpueh 

I think a summary of what you mean is that `DELETE` rule should be included in materials rules and `CREATE` rule should be included in products rules?

I understand your point about the `DELETE` rule now. I am also thinking that is `DISALLOW *` in products rules do a similar job as `DELETE` rule? Since `DISALLOW *` indicates that all other artifacts are not allowed as products, does it also indicates that deleted artifacts should not show up in the products queue?

lukpueh replied:
> I think a summary of what you mean is that DELETE rule should be included in materials rules and CREATE rule should be included in products rules?

Exactly! :)

> I am also thinking that is DISALLOW * in products rules do a similar job as DELETE rule?

They are actually quite different. DELETE  is a rule that consumes artifacts, i.e. removes it from the queue when matched by the pattern. Since deleted files can only be in the materials queue, which is only being looked at when processing material rules, they can also only be consumed there.
The DISALLOW rule, on the other hand, does not consume artifacts. Instead it fails the verification if its pattern matches any artifacts in the corresponding queue (material queue if the DISALLOW rule is in expected_materials, and product queue if the DISALLOW rule is in the expected_products).

lukpueh reviewed (commented):
Excellent work, @SolidifiedRay and @IshaDave! 💯 for clean code, extensive tests and a very good first stab at the rule generation.

Maybe you two can start sketching some ideas how to add match rules by looking at  multiple consecutive links?

lukpueh commented on tests/test_before_after_filesystem_snapshot.py:
```diff
@@ -0,0 +1,113 @@
+    }
+    snapshot = before_after_filesystem_snapshot.snapshot(self.before, after)
+    rules = before_after_filesystem_snapshot.generate_artifact_rules(snapshot)
+    self.assertEqual(sorted(artifact_rules), sorted(rules))
```
I doubt that this is what you want to test. See [my comment about calling sorted on a dict](https://github.com/in-toto/layout-web-tool/pull/46#discussion_r464386252). You probably `want assertDictEqual` instead.

SolidifiedRay replied:
Oh sorry, I somehow miss your previous comment, but now I understand why I shouldn't sort the dictionary. Thanks!

lukpueh commented on before_after_filesystem_snapshot.py:
```diff
@@ -0,0 +1,65 @@
+  expected_materials = []
+  expected_products = []
+
+  # missing rules for MATCH since we don't have the information of the material from 
```
Let's start this comment with `# TODO: ` and take care of it in a follow up PR.

IshaVDave replied:
Thank you for so many reviews Lukas! I added `#TODO: ` and the Travis check runs and passes. I think it is ready to merge now. 

> Could you please also create a new issue that very briefly and roughly describes what we want to do next (read in link metadata, generate match rules that span multiple links, etc...)?

Sure.

lukpueh reviewed (changes requested):
Just gave it another pass and ran the tests locally. This is solid work, @IshaDave and @SolidifiedRay! 💯  Let's merge and add new features in follow up PRs.

FYI, I just enabled Travis for this repo, so if you do another push (e.g. to update the code comment as suggested below) Travis should run the tests.

Could you please also create a new issue that very briefly and roughly describes what we want to do next (read in link metadata, generate match rules that span multiple links, etc...)? We can then use that issue page to discuss further ideas.  

Thanks!

SolidifiedRay committed:
Add function to generate artifact rules.

Adds a function to before_after_filesystem_snapshot.py called
generate_artifact_rules.

lukpueh reviewed (approved):
LGTM! let’s merge here and keep the discussion going in #48. :)
